### ISSUE TITLE
Failure on kubernetes due to walking hidden subdirectories

### ISSUE THREAD

**Issue by BenPope** [lyft/goruntime#15]
I've tried to deploy https://github.com/lyft/ratelimit to Kubernetes, but it fails to find the correct configuration because goruntime first traverses a hidden directory, and then fails with a duplicate domain on the configuration it should actually load.

When Kubernetes mounts a configMap, it does symlink trickery to allow for atomic swapping; e.g.,:
```
bash-4.4$ ls -al 
total 32
drwxrwxrwx 3 root root 4096 May  9 17:03 .
drwxr-xr-x 3 root root 4096 May  9 17:03 ..
drwxr-xr-x 2 root root 4096 May  9 17:03 ..2018_05_09_17_03_25.201237914
lrwxrwxrwx 1 root root   31 May  9 17:03 ..data -> ..2018_05_09_17_03_25.201237914
lrwxrwxrwx 1 root root   18 May  9 17:03 config.yaml -> ..data/config.yaml
bash-4.4$ ls -al ..2018_05_09_17_03_25.201237914 
total 24
drwxr-xr-x 2 root root 4096 May  9 17:03 .
drwxrwxrwx 3 root root 4096 May  9 17:03 ..
-rw-r--r-- 1 root root  210 May  9 17:03 config.yaml
```
goruntime first loads `..2018_05_09_17_03_25.201237914/config.yaml`, and then `./config.yaml` is rejected with a duplicate `domain`.  Due to the way goruntime forms a hierarchical key, ratelimit is unable to find suitable configuration.

Would it make sense to prevent walking directories that start with a dot?  I guess they don't make a lot of sense given that the hierarchical key separator is also a dot.

I'm happy to make a PR if there is general agreement.

**Comment by junr03** [comment]
@BenPope thanks for the report. @lemccarthy just brought this up to my attention. @theatrus @paulnivin  do you mind giving your opinion on the proposed solution?

**Comment by theatrus** [comment]
Avoiding hidden (dot) directories sounds like a good solution overall. Our source for configuration internally is not a configmap (for various reasons) so we haven’t run into this before on Kubernetes.

**Comment by junr03** [comment]
@BenPope do you mind opening a PR against this repo to fix the issue like you have done in your fork. Thanks

**Comment by junr03** [comment]
Additionally, I would propose making the feature configurable.

**Comment by lemccarthy** [comment]
I tested the fix on his PR on his fork, and I'm not guaranteeing I did it correctly, but while it eliminated one of my error messages (where it tried to read `..data` and complained it was a directory), I still get the duplicate key message.

You can look at the fork I made: https://github.com/lemccarthy/goruntime

**Comment by BenPope** [comment]
@lemccarthy is the duplicated key message to do with the domain in ratelimit?

**Comment by lemccarthy** [comment]
@BenPope Yes, it claims there is a duplicate domain, even though my config file has only one domain.

**Comment by BenPope** [comment]
How odd.  This is the branch of ratelimit I've been using: https://github.com/pusher/lyft-ratelimit/tree/ignore-dotfiles.

The test I wrote above passes against: https://github.com/lyft/goruntime/pull/17

**Comment by lemccarthy** [comment]
I'm entirely willing to grant that I messed it up. It could also possibly be something in the way I'm doing my kubernetes setup or my docker build (I have to build a docker image to use in my kubernetes pod). The file is definitely there and being read, its just clearly being read multiple times.

I will say that your code fixed *one* error I was having. Previously it explicitly complained about `../data` saying it could not read a directory. Now it simply complains about the duplicate key, so the code additions had a nonzero effect.

I'll try switching the glide depend to your fork and see if that does it.

**Comment by lemccarthy** [comment]
@BenPope I cloned that branch, and it fails on "glide up" claiming it can't figure out where to find the goruntime repo.

**Comment by lemccarthy** [comment]
@BenPope if you'll view my edit, it is now working on my kubernetes setup! I just had to change that line in `glide.yaml` under the "repo" tag for the goruntime dependency from `git@github.com:pusher/lyft-goruntime` to "https://github.com/pusher/lyft-goruntime

I have the deployment up and running now with a working helmchart
```
kubectl logs ratelimiter-0 -c ratelimit
time="2018-09-14T14:10:09Z" level=warning msg="statsd is not in use"
time="2018-09-14T14:10:09Z" level=warning msg="connecting to redis on tcp 127.0.0.1:6379 with pool size10"
time="2018-09-14T14:10:09Z" level=warning msg="Listening for HTTP on ':8080'"
time="2018-09-14T14:10:09Z" level=warning msg="Listening for debug on ':6070'"
time="2018-09-14T14:10:09Z" level=warning msg="Listening for gRPC on ':8081'"
kubectl exec -it ratelimiter-0 -c ratelimit -- curl 0:6070/rlconfig
foo.database_users: unit=SECOND requests_per_unit=500
```

**Comment by lemccarthy** [comment]
@BenPope Sorry to bother you again, but although the service is working/loading a config, I'm having trouble actually routing stuff to it. Are you using Ambassador with ratelimit in your stack?

**Comment by BenPope** [comment]
@lemccarthy I'm calling into it from a bespoke product, unfortunately I haven't tested it with envoy or ambassador rate limiting.

**Comment by lemccarthy** [comment]
@BenPope ah that's a shame. Thanks for getting back to me.

**Comment by BenPope** [comment]
@junr03 - did you get a chance to see #17? Happy to make changes.

**Comment by junr03** [comment]
oh, no sorry didn't see that getting posted. I will assign myself so that it is my queue of oss reviews for tomorrow. Thanks
